src/Type/Unify.hs

```diff
@@ -267,8 +267,8 @@ unify (TApp t1 ts1) (TApp u1 us2)   -- | length ts1 != length us2
                 unifies (drop (len1 - len2) ts1) us2
 
 -- functions
-unify f1@(TFun args1 eff1 res1) f2@(TFun args2 eff2 res2) | length args1 == length args2
-  = do unifies (res1:map snd args1) (res2:map snd args2)
+unify f1@(TFun args1 eff1 res1) f2@(TFun args2 eff2 res2) | length (takeNonImplicits args1) == length (takeNonImplicits args2)
+  = do unifies (res1:map snd (takeNonImplicits args1)) (res2:map snd (takeNonImplicits args2))
```

Comment:
I'm not sure if this is correct for first-class / higher order implicits. We probably want to take the min length of the two lists.